### ISSUE TITLE
torch.cuda.amp.autocast nan values when model.eval()

### ISSUE THREAD

**Issue by ADulian** [pytorch/pytorch#58392]
Hi there,

I seem to have an issue with training on mixed precision with torch.cuda.amp.autocast, during training loop i.e. when net.train() the network outputs is valid, however, changing the mode to net.eval() causes some of the data samples to produce NaN values when ran through net.forward(). This issue is not present on every epoch, however, it is present for most of the training. I checked if there is an issue with the datasets by running validation/test loaders during training and by running train loader during evaluation. It seems that the issue is only present when the model is set to eval(), during train() everything seems to be absolutely fine. 

Greatly appreciate any advice on how to solve this issue and what can be causing it.

Cheers

**Comment by heitorschueroff** [comment]
@ADulian Could you share some code snippet that reproduces the problem you're describing as well as what the expected behavior should be. Any extra information will help us help you better. Thank you!

**Comment by ADulian** [comment]
@heitorschueroff thanks for getting back to me!

If I execute the following
`self.model.eval(), out = self.model(data)`

then out will contain NaN values, the tensor is not randomly filled with just some NaNs but rather all values within it are NaNs. However, if instead I do

`self.model.train(), out = self.model(data)`

then the `out` tensor outputs expected values (for the same data sample). 

During training the NaN do not appear on every epoch, e.g. the following shows the log file from 4 training epochs:

Epoch: 43/100.....26808
Training:   Loss: 2.41e-03  
Validation: Loss: nan    

--- Best Validation Loss: 2.47e-03 Epoch: 42 File: (13_50_21)_(16_05_2021) ---

Epoch: 44/100.....26808
Training:   Loss: 2.29e-03    
Validation: Loss: 2.18e-03    

--- Validation Loss improved from: 2.47e-03 to: 2.18e-03

Epoch: 45/100.....26808
Training:   Loss: 2.13e-03   
Validation: Loss: nan    

--- Best Validation Loss: 2.18e-03 Epoch: 44 File: (13_50_21)_(16_05_2021) ---

Epoch: 46/100.....26808
Training:   Loss: 2.01e-03
Validation: Loss: nan   

--- Best Validation Loss: 2.18e-03 Epoch: 44 File: (13_50_21)_(16_05_2021) ---

I do wonder why changing from `self.model.eval()` to `self.model.train()` fixes the issue. In both cases the data sample when testing was the same with the `dtype=fp16` and the `torch.cuda.amp.autocast(enabled=True)` . 

The model that I'm using is an implementation of the following:
[https://github.com/HRNet/HigherHRNet-Human-Pose-Estimation](url)
[https://github.com/HRNet/HigherHRNet-Human-Pose-Estimation/blob/master/lib/models/pose_higher_hrnet.py](url)

**Comment by StaudTim** [comment]
Have you found a solution? I'm facing the same problem, but with the difference that the loss is NaN in every epoch.

**Comment by YudiXie** [comment]
I am facing the same issue, when I train the pytroch ResNet-50 model on an image dataset.
Before using AMP, both the training and validation were fine. After using autocast, validation occasionally gives me nan values, but the training is fine. This only happens to some models but don't happens to other models with different random seeds. Here is a part of my training log.

```
Pytorch version:  2.0.1
Using GPU, Pytorch linked cuda version:  11.7
Loaded checkpoint! Restarting from batch: 246000
Batch Number:     260000, Train Loss: 0.012, Valid Loss: 0.201
Batch Number:     280000, Train Loss: 0.016, Valid Loss: 0.215
Batch Number:     300000, Train Loss: 0.008, Valid Loss: nan
Batch Number:     320000, Train Loss: 0.007, Valid Loss: 0.202
Batch Number:     340000, Train Loss: 0.011, Valid Loss: nan
Batch Number:     360000, Train Loss: 0.007, Valid Loss: 0.204
Batch Number:     380000, Train Loss: 0.006, Valid Loss: 0.205
Batch Number:     400000, Train Loss: 0.019, Valid Loss: nan
Batch Number:     420000, Train Loss: 0.008, Valid Loss: 0.205
Batch Number:     440000, Train Loss: 0.005, Valid Loss: 0.206
Batch Number:     460000, Train Loss: 0.005, Valid Loss: 0.203
Batch Number:     480000, Train Loss: 0.005, Valid Loss: 0.200
Batch Number:     500000, Train Loss: 0.007, Valid Loss: nan
Batch Number:     520000, Train Loss: 0.009, Valid Loss: 0.214
Batch Number:     540000, Train Loss: 0.004, Valid Loss: nan
Batch Number:     560000, Train Loss: 0.006, Valid Loss: nan
Batch Number:     580000, Train Loss: 0.009, Valid Loss: nan
Batch Number:     600000, Train Loss: 0.006, Valid Loss: nan
Batch Number:     620000, Train Loss: 0.008, Valid Loss: 0.228
Batch Number:     640000, Train Loss: 0.005, Valid Loss: 0.216
Batch Number:     660000, Train Loss: 0.004, Valid Loss: nan
Batch Number:     680000, Train Loss: 0.005, Valid Loss: 0.210
Batch Number:     700000, Train Loss: 0.002, Valid Loss: nan
Batch Number:     720000, Train Loss: 0.007, Valid Loss: nan
Batch Number:     740000, Train Loss: 0.004, Valid Loss: nan
Batch Number:     760000, Train Loss: 0.002, Valid Loss: nan
Batch Number:     780000, Train Loss: 0.005, Valid Loss: nan
```
